### ISSUE TITLE
Hilt Feature Request: Missing compile classpath dep for optional binding should be compile error

### ISSUE THREAD

**Issue by snepalnetflix** [google/dagger#2223]
If a feature is optional (via `@BindsOptionalOf`) and accessed in a feature module using `@EntryPoint`, then the feature can be missing from the hilt component but no compile error is generated. Instead, accessing the feature results in a runtime crash.

To recreate this problem you need 3 modules:

1. app (root module)
2. home (module that uses the feature)
3. feature (declares the optional feature and the `@EntryPoint` accessor)

In this case, app depends on home, and home depends on feature. Because feature is not in the compile classpath of app, dagger can't validate the graph.

Reproduction steps:
```
git clone git@github.com:snepalnetflix/hilt-optional-feature.git
cd hilt-optional-feature
./gradlew :app:assembleDebug
adb install -r -d ./app/build/outputs/apk/debug/app-debug.apk
adb shell monkey -p com.example.daggerruntimeclasspatherror -c android.intent.category.LAUNCHER 1
```
Result:
App crashes with the following exception:
```
java.lang.ClassCastException: Cannot cast com.example.daggerruntimeclasspatherror.DaggerMyApplication_HiltComponents_SingletonC to com.example.feature.Feature$FeatureSingletonAccessor
```

Note, if you change the code to reference the dependency in the home module, the error changes from a runtime crash to a compile error. To see this, uncomment the commented out code in `home/src/main/java/com/example/home/MainActivity.kt`. Then do:
```
./gradlew clean
./gradlew :app:assembleDebug
```
You get the following compile error:
```
error: cannot access Feature
  class file for com.example.feature.Feature not found
```

Note, to fix the above compile error you would edit app/build.gradle and add `implementation project(":feature")`.

**Comment by snepalnetflix** [comment]
Something completely unrelated, in the above reproduction steps, if you don't perform `./gradlew clean`, the compilation succeeds. But if you do perform `./gradlew clean`, then compilation fails. I think this is indicative of another bug.

**Comment by bcorso** [comment]
Hi @snepalnetflix,

I believe this is a dupe of #2139 and #1991.

@danysantiago is currently looking into a fix, which is being tracked in #1991.

**Comment by Chang-Eric** [comment]
Going to close this since it looks like it is a dupe of the issues mentioned above. Please let us know if you disagree though and we can take another look.
